### PR TITLE
[record-minmax] Revise to prepare to multiple instance

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This pr revises record-minmax to prepare to using multiple instance of luci-interpreters and observers.

for issue https://github.com/Samsung/ONE/issues/10133
draft: #10132
first step according to https://github.com/Samsung/ONE/pull/10132#discussion_r1034138863

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

### CODE REVIEW

File: compiler/record-minmax/CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,16 @@ target_link_libraries(record-minmax dio_hdf5)
 target_link_libraries(record-minmax vconone)
 target_link_libraries(record-minmax nncc_coverage)
 
+find_package(OpenMP QUIET)
+if (OpenMP_FOUND)
+  set(_OPENMP ON)
```
Comment (seanshpark) [inline compiler/record-minmax/CMakeLists.txt:20]:
```suggestion
  set(RECORDMINMAX_OPENMP ON)
```

Comment (seanshpark) [reply compiler/record-minmax/CMakeLists.txt:20]: `_OPENMP` is generic and it hard to identify where this came from.

Comment (seanshpark) [reply compiler/record-minmax/CMakeLists.txt:20]: Q) what is the purpose of `_OPENMP` set to `ON`?

File: compiler/record-minmax/include/RecordMinMax.h [hunk 2]
```diff
@@ -24,13 +24,23 @@
 
 #include <memory>
 
+#ifdef _OPENMP
+#include <omp.h>
```
Comment (seanshpark) [inline compiler/record-minmax/include/RecordMinMax.h:28]:
```suggestion
// refer https://www.openmp.org/spec-html/5.0/openmpse10.html for _OPENMP
#ifdef _OPENMP
#include <omp.h>
```

**Comment by jinevening** [comment]
I'll review this PR after https://github.com/Samsung/ONE/issues/10133#issuecomment-1331508157 is answered.

File: compiler/record-minmax/CMakeLists.txt (hunk 1 above)
Comment (BalyshevArtem) [reply compiler/record-minmax/CMakeLists.txt:20]: Actually, for clarity, but if it's unnecessary, I can remove :)

Comment (seanshpark) [reply compiler/record-minmax/CMakeLists.txt:20]: OK, I think `_OPENMP` is defined from `OpenMP ` package. I you agree, we can remove this.

**Review by SlavikMIPT — DISMISSED** [review]
LGTM

File: compiler/record-minmax/include/RecordMinMax.h [hunk 3]
```diff
@@ -51,9 +61,19 @@ class RecordMinMax
   void saveModel(const std::string &output_model_path);
 
 private:
+  void initializeCircleModule(const std::string &input_model_path);
+
+  luci_interpreter::Interpreter *getInterpreter() { return _interpreters[0].get(); }
```
Comment (jinevening) [inline compiler/record-minmax/include/RecordMinMax.h:66]:
```suggestion
  luci_interpreter::Interpreter *getInterpreter() const { return _interpreters[0].get(); }
```

File: compiler/record-minmax/include/RecordMinMax.h [hunk 4]
```diff
@@ -51,9 +61,19 @@ class RecordMinMax
...
+
+  MinMaxObserver *getObserver() { return _observers[0].get(); }
```
Comment (jinevening) [inline compiler/record-minmax/include/RecordMinMax.h:68]:
```suggestion
  MinMaxObserver *getObserver() const { return _observers[0].get(); }
```

File: compiler/record-minmax/include/RecordMinMax.h [hunk 5]
```diff
@@ -51,9 +61,19 @@ class RecordMinMax
...
+
   std::unique_ptr<luci::Module> _module;
-  std::unique_ptr<luci_interpreter::Interpreter> _interpreter;
-  std::unique_ptr<MinMaxObserver> _observer;
+
+  // Multiple interpreters are used for parallel execution
+  std::vector<std::unique_ptr<luci_interpreter::Interpreter>> _interpreters;
+  std::vector<std::unique_ptr<MinMaxObserver>> _observers;
+
+  uint32_t _threads_size;
```
Comment (jinevening) [inline compiler/record-minmax/include/RecordMinMax.h:76]: For our SE score.

```suggestion
  uint32_t _threads_size = 0;
```

File: compiler/record-minmax/src/RecordMinMax.cpp [hunk 6]
```diff
@@ -200,13 +200,26 @@ void RecordMinMax::initialize(const std::string &input_model_path)
...
+}
 
-  // Initialize interpreter
-  _interpreter = std::make_unique<luci_interpreter::Interpreter>(_module.get());
+void RecordMinMax::initialize(const std::string &input_model_path)
+{
+  initializeCircleModule(input_model_path);
 
-  _observer = std::make_unique<MinMaxObserver>();
+  // Create and initialize interpreters and observers
```
Comment (jinevening) [inline compiler/record-minmax/src/RecordMinMax.cpp:209]:
```suggestion
  assert(_threads_size > 0);

  // Create and initialize interpreters and observers
```

File: compiler/record-minmax/include/RecordMinMax.h [hunk 7]
```diff
@@ -24,13 +24,23 @@
...
 namespace record_minmax
 {
 
 class RecordMinMax
 {
 public:
-  explicit RecordMinMax() = default;
+  explicit RecordMinMax(bool use_parallel) : _threads_size(use_parallel ? omp_get_max_threads() : 1)
+  {
+    // Do nothing
```
Comment (jinevening) [inline compiler/record-minmax/include/RecordMinMax.h:42]:
```suggestion
    assert(_threads_size > 0);
```

Comment (BalyshevArtem) [reply compiler/record-minmax/include/RecordMinMax.h]: Done

**Comment by BalyshevArtem** [comment]
updated and rebased this pr

File: compiler/record-minmax/src/RecordMinMax.cpp [hunk 8]
```diff
@@ -169,7 +169,7 @@ void update_quantparam(record_minmax::MinMaxObserver *observer, const std::strin
 namespace record_minmax
 {
 
-void RecordMinMax::initialize(const std::string &input_model_path)
+void RecordMinMax::initializeCircleModule(const std::string &input_model_path)
```
Comment (seanshpark) [inline compiler/record-minmax/src/RecordMinMax.cpp:172]: can you please make this diff better to review?
unintended changes are visible and I cannot catch the changes.

Comment (BalyshevArtem) [reply compiler/record-minmax/src/RecordMinMax.cpp]: Done

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

File: compiler/record-minmax/src/RecordMinMax.cpp [hunk 9]
```diff
@@ -171,6 +171,8 @@ namespace record_minmax
 
 void RecordMinMax::initialize(const std::string &input_model_path)
 {
+  assert(_threads_size > 0);
+
```
Comment (jinevening) [inline compiler/record-minmax/src/RecordMinMax.cpp:175]: ditto

File: compiler/record-minmax/driver/Driver.cpp [hunk 10]
```diff
@@ -92,6 +92,7 @@ int entry(const int argc, char **argv)
   float min_percentile = 1.0;
   float max_percentile = 99.0;
   std::string input_data_format("h5");
+  int num_threads = 1;
```
Comment (jinevening) [inline compiler/record-minmax/driver/Driver.cpp:95]:
```suggestion
  uint32_t num_threads = 1;
```

File: compiler/record-minmax/include/RecordMinMax.h [hunk 11]
```diff
@@ -23,14 +23,18 @@
...
+#include <thread>
 
 namespace record_minmax
 {
 
 class RecordMinMax
 {
 public:
-  explicit RecordMinMax() = default;
+  explicit RecordMinMax(int32_t num_threads) : _threads_size(num_threads)
```
Comment (jinevening) [inline compiler/record-minmax/include/RecordMinMax.h:34]: Our SE tool may complain implicit type conversion int32 -> uint32. Could you change dtype of num_threads to `uint32_t`?

File: compiler/record-minmax/driver/Driver.cpp (hunk 10 above)
Comment (jinevening) [inline compiler/record-minmax/driver/Driver.cpp:95]:
```suggestion
  uint32_t num_threads = 1;
```

**Comment by seanshpark** [comment]
@jinevening , please apply this for `int` vs `uint` compares (in another issue, another PR)
```diff

index b9c08f472..b93e06dc0 100644
--- a/compiler/record-minmax/CMakeLists.txt
+++ b/compiler/record-minmax/CMakeLists.txt
@@ -14,6 +14,7 @@ target_link_libraries(record-minmax luci_interpreter)
 target_link_libraries(record-minmax dio_hdf5)
 target_link_libraries(record-minmax vconone)
 target_link_libraries(record-minmax nncc_coverage)
+target_link_libraries(record-minmax nncc_common)

 install(TARGETS record-minmax DESTINATION bin)

```

**Review by jinevening — APPROVED** [review]

**Pull request merged**